src/components/cuda_pw/linux-cuda.c

```diff
@@ -0,0 +1,374 @@
+    _cuda_pw_vector.cmp_info.num_native_events = -1;
+    _cuda_pw_vector.cmp_info.num_cntrs = -1;
+    _cuda_pw_lock = PAPI_NUM_LOCK + NUM_INNER_LOCK + cidx;
+
+    _cuda_pw_vector.cmp_info.initialized = 1;
+    _cuda_pw_vector.cmp_info.disabled = PAPI_EDELAY_INIT;
+    sprintf(_cuda_pw_vector.cmp_info.disabled_reason,
+        "Not initialized. Access component events to initialize it.");
+    return PAPI_EDELAY_INIT;
+}
+
+static int cuda_shutdown_component(void)
+{
+    COMPDBG("Entering.\n");
+    int res = PAPI_OK;
+    free_event_name_list(&global_event_names);  // Free the global event names added dynamically
+
+    if (!_cuda_pw_vector.cmp_info.initialized ||
+    _cuda_pw_vector.cmp_info.disabled != PAPI_OK)
+    return PAPI_OK;
+
+    _cuda_pw_vector.cmp_info.initialized = 0;
+
+    // res = cupti_shutdown();
+    // Reverse of private init.
+    return res;
+}
+
+static int cuda_init_private(void)
+{
+    int res = PAPI_OK;
+    const char *disabled_reason;
+    COMPDBG("Entering.\n");
+    res = cupti_init(&disabled_reason);
+    if (res != PAPI_OK) {
+        goto fn_exit;
+    }
+
+    // Initialize global_event_names array
+    res = initialize_dynamic_event_list(&global_event_names);
+    if (res != PAPI_OK) {
+        goto fn_exit;
+    }
+
+    _cuda_pw_vector.cmp_info.disabled = PAPI_OK;
+    strcpy(_cuda_pw_vector.cmp_info.disabled_reason, "");
+
+fn_exit:
+    return res;
+}
+
+static int check_n_initialize(void)
+{
+    _papi_hwi_lock(COMPONENT_LOCK);
+    int res = PAPI_OK;
+    if (_cuda_pw_vector.cmp_info.initialized
+        && _cuda_pw_vector.cmp_info.disabled == PAPI_EDELAY_INIT
+    ) {
+        res = cuda_init_private();
+    }
+
+    _papi_hwi_unlock(COMPONENT_LOCK);
+    return res;
+}
+
+static int cuda_ntv_enum_events(unsigned int __attribute__((unused)) *event_code, int __attribute__((unused)) modifier)
+{
+
+    int res = check_n_initialize();
+    if (res != PAPI_OK)
+        return PAPI_ENOEVNT;  // Not implemented
+    else
+        return PAPI_OK;
+}
+
+static int cuda_ntv_name_to_code(const char *name, unsigned int *event_code)
+{
+    int res = check_n_initialize();
+    if (res != PAPI_OK)
+        return res;
+    event_rec_t *evt_rec;
+    res = find_event_name(&global_event_names, name, &evt_rec);
+    if (res == PAPI_OK) {
+        *event_code = evt_rec->evt_code;
+        return PAPI_OK;
+    }
+    else {
+        _papi_hwi_lock(COMPONENT_LOCK);
+        *event_code = global_event_names.count;
+        res = insert_event_record(&global_event_names, name, global_event_names.count, 0);
+        _papi_hwi_unlock(COMPONENT_LOCK);
+    }
+
+    return PAPI_OK;
+}
+
+static int cuda_ntv_code_to_name(unsigned int event_code, char *name, int len)
+{
+    int res = check_n_initialize();
+    if (res != PAPI_OK)
+        return res;
+    if (event_code >= global_event_names.count) {
+        return PAPI_EINVAL;
+    }
+
+    strncpy(name, global_event_names.evts[event_code].name, len);
```

Comment:
Same comment as above. First search in frontend maintained event table and, if look up fails, call `cupti_evt_code_to_name()` from the backend.

Reply:
Please understand the design first. There is no back end maintained table. There is only the `event_list_t global_event_names`. Names are added dynamically to this table when `cuda_ntv_name_to_code` is called with a new name, and an event code is assigned to this name, it returns the new event code or the existing code. `cuda_ntv_code_to_name` will look in this global event table, if it doesn't find it, it should return `PAPI_ENOEVNT` (currently returns PAPI_EINVAL).

When `cuda_ntv_enum_events` is called, it should populate the global table so that the upper layer event name <-> code api works as expected.

Reply:
Correct. That's how it should work. Let me make an implementation and you can comment on it.

Reply:
Have a look at commit [cuda_pw: enumerate event names implemented. but no descriptions.](https://github.com/anustuvicl/papi/pull/1/commits/78da4c99a1fd24f244cdf43a3a21aa4b67944ea1)